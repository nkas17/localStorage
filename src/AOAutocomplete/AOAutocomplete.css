.arrf-autocomplete.arrf-autocomplete.arrf-autocomplete {
  z-index: 90000;
  padding: 0;
  position: absolute;
  top: 32px;
  left: 0;
  background: rgba(255, 255, 255, 0.9);
  border: solid 1px #ccc;
  border-radius: 3px;
  box-shadow: 0 2px 12px rgba(0, 0, 0, 0.1);
  font-size: 90%;
  overflow: auto;
}

.arrf-autocomplete__item.arrf-autocomplete__item.arrf-autocomplete__item {
 font-size: 1rem;
 cursor: default;
padding: 2px 6px;
}

.arrf-autocomplete__item--highlighted.arrf-autocomplete__item--highlighted.arrf-autocomplete__item--highlighted {
  color: white;
  background: hsl(200, 50%, 50%);
}
