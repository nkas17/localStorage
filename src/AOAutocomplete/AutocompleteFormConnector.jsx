import _ from 'lodash';
import React from 'react';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
import Autocomplete from 'react-autocomplete';

/**
 * Renders the react-autocomplete
 * This component should be invoked by the redux-form Field component
 *
 * @component AutocompleteFormConnector
 * @import AutocompleteFormConnector
 * @see https://github.com/reactjs/react-autocomplete/blob/master/lib/Autocomplete.js
 */
/* eslint-disable react/prop-types */
const AutocompleteFormConnector = (props) => {
	const optionalAttributes = {};
	if (!_.isNil(props.open)) optionalAttributes.open = props.open;
	return (
		<ReactCSSTransitionGroup
			transitionName="onShow"
			transitionAppear={props.transition}
			transitionAppearTimeout={500}
			transitionEnter={false}
			transitionLeave={false}
			transitionEnterTimeout={500}
			transitionLeaveTimeout={500}
		>
			<Autocomplete
				inputProps={
					Object.assign({}, {
						className: `ao-textbox ${props.widthClassName} ${props.additionalClassNames} ${props.meta.touched && props.meta.error !== undefined
							? 'ao-error' : ''} ${props.meta.touched && props.meta.warning !== undefined ? 'ao-warning' : ''}`,
						name: props.input.name,
						id: props.id,
						onBlur: ({
							target: {
								value: caption,
							},
						}) => {
							const value = props.getValueFromCaption(caption, props.items);
							return props.input.onBlur(value);
						},
						onDragStart: props.input.onDragStart,
						onDrop: props.input.onDrop,
						onFocus: props.input.onFocus,
						placeholder: props.placeholder,
					}, props.inputProps)
				}
				value={props.getCaptionFromValue(props.input.value, props.items)}
				getItemValue={props.getItemValue}
				items={props.items}
				renderItem={props.renderItem || props.renderItemGenerator(props.getCaptionFromValue(props.input.value, props.items))}
				onChange={({
					target: {
						value: caption,
					},
				}) => {
					const value = props.getValueFromCaption(caption, props.items);
					return props.input.onChange(value);
				}}
				onSelect={(caption) => {
					const value = props.getValueFromCaption(caption, props.items);
					props.input.onChange(value);
					return props.onSelect(value);
				}}
				shouldItemRender={props.shouldItemRender}
				sortItems={props.sortItems}
				autoHighlight={props.autoHighlight}
				onMenuVisibilityChange={props.onMenuVisibilityChange}
				renderMenu={props.renderMenu}
				wrapperProps={props.wrapperProps}
				wrapperStyle={props.wrapperStyle}
				{...optionalAttributes}
			>
				{props.children}
			</Autocomplete>
		</ReactCSSTransitionGroup>
	);
};

export default AutocompleteFormConnector;
