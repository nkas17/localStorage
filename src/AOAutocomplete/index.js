import _AOAutocomplete from './AOAutocomplete';

export * from './utility';
export const AOAutocomplete = _AOAutocomplete;
