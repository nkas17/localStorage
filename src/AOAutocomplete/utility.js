import React from 'react';
import _ from 'lodash';

export const getItemValue = item => item.caption;

// Unique key used to make elements unique when looping.
let key = 0;

/**
 * Renders the caption so that substrings matching what the user typed are bold
 * http://regexr.com/3g8tg
 *
 * @param {string} caption - string to render
 * @returns {Node[]} - array of strings, potentially with bold tags around some
 */
export const renderCaption = (caption, inputValue) => {
	const escapedInputValue = inputValue.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&');
	const findInputRegex = new RegExp(`(^${escapedInputValue})|(\\W)(${escapedInputValue})`, 'ig');
	const splitCaption = caption.split(findInputRegex);
	return _.reduce(splitCaption, (elementsSoFar, subCaption) => {
		if (_.isNil(subCaption)) {
			return elementsSoFar;
		}
		if (subCaption.toLowerCase() === inputValue.toLowerCase()) {
			return elementsSoFar.concat(<b key={++key}>{subCaption}</b>);
		}
		return elementsSoFar.concat(subCaption);
	}, []);
};

/**
 * Creates the default renderItem function which bolds
 * the parts of the caption matching inputValue.
 *
 * We use a higher-order function because we do not have control of the "input" in this scope
 *
 * @param {string} inputValue - value on the input tag
 */
export const renderItemGenerator = inputValue => (item, isHighlighted) => (
	<div
		key={++key}
		className={`arrf-autocomplete__item ${isHighlighted ? 'arrf-autocomplete__item--highlighted' : ''}`}
	>
		{renderCaption(item.caption, inputValue)}
	</div>
);

/**
 * Transforms the string to filter out unwanted formatting.
 * Currently changes the string to be lower case and replaces punctuation with spaces
 * @param {string} string
 * @returns {string} - the transformed string
 */
const _transformString = (string) => {
	const lowerCase = string.toLowerCase();
	const withoutPunctuation = lowerCase.replace(/[[\].,/#!$%^&*;:{}=\-_`~()\\]/g, ' ');
	return withoutPunctuation;
};
// / ^ *
// Items will appear in the menu if the value entered matches the beginning of a word from the caption or value
export const shouldItemRender = (item, currentValue) => {
	const transformedItemValue = _transformString(item.value);
	const transformedItemCaption = _transformString(item.caption);
	const transfoemdCurrentValue = _transformString(currentValue);
	return ` ${transformedItemValue} ${transformedItemCaption}`.toLowerCase().includes(` ${transfoemdCurrentValue}`);
};

// sort items by match position by default
// see https://github.com/reactjs/react-autocomplete/blob/master/lib/utils.js
export const sortItems = (a, b, value) => {
	const aLower = a.caption.toLowerCase();
	const bLower = b.caption.toLowerCase();
	const valueLower = value.toLowerCase();
	const queryPosA = aLower.indexOf(valueLower);
	const queryPosB = bLower.indexOf(valueLower);
	if (queryPosA !== queryPosB) {
		return queryPosA - queryPosB;
	}
	return aLower < bLower ? -1 : 1;
};

export const renderMenu = items => (
	<div className="arrf-autocomplete">{items}</div>
);

/**
 * Returns the caption to the item whose value property matches value.
 * If none are found returns the original value.
 *
 * @param {string} value - value to search for in the items
 * @param {object[]} items - array of items to query
 * @returns {string} - the caption to display
 */
export const getCaptionFromValue = (value, items) => {
	const itemWithValue = items.find(item => item.value === value);
	if (itemWithValue) {
		return itemWithValue.caption;
	}
	return value;
};

/**
 * Returns the value to the item whose caption property matches caption.
 * If none are found returns the original caption.
 *
 * @param {string} caption - caption to search for in the items
 * @param {object[]} items - array of items to query
 * @returns {string} - the value to store
 */
export const getValueFromCaption = (caption, items) => {
	const itemWithCaption = items.find(item => item.caption === caption);
	if (itemWithCaption) {
		return itemWithCaption.value;
	}
	return caption;
};
