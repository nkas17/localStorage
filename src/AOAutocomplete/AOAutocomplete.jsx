import React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'redux-form';
import AutocompleteFormConnector from './AutocompleteFormConnector';
import * as defaults from './utility';
import './aoAutocomplete.css';

/**
 * Creates an autocomplete component.
 * To customize the menu styles, instead of using menuStyle, customize the css class .arrf-autocomplete
 * To customize the item styles, use the css class .arrf-autocomplete__item
 * To customize the highlighted item styles, use the css class .arrf-autocomplete__item--highlighted
 *
 * @component AOAutocomplete
 * @import AOAutocomplete
 * @see https://github.com/reactjs/react-autocomplete/blob/master/lib/Autocomplete.js
 */
const AOAutocomplete = ({
	getItemValue,
	items,
	renderItem,
	renderItemGenerator,
	shouldItemRender,
	sortItems,
	onBlur,
	onChange,
	onDragStart,
	onDrop,
	onFocus,
	id,
	name,
	widthClassName,
	additionalClassNames,
	rightLabel,
	transition,
	placeholder,
	validate,
	warn,
	children,
	autoHighlight,
	inputProps,
	onMenuVisibilityChange,
	onSelect,
	renderMenu,
	wrapperProps,
	wrapperStyle,
	open,
	getCaptionFromValue,
	getValueFromCaption,
}) => (
	<Field
		component={AutocompleteFormConnector}
		onBlur={onBlur}
		onChange={onChange}
		onDragStart={onDragStart}
		onDrop={onDrop}
		onFocus={onFocus}
		id={id}
		name={name}
		widthClassName={widthClassName}
		additionalClassNames={additionalClassNames}
		rightLabel={rightLabel}
		transition={transition}
		placeholder={placeholder}
		validate={validate}
		warn={warn}
		getItemValue={getItemValue}
		items={items}
		renderItem={renderItem}
		shouldItemRender={shouldItemRender}
		sortItems={sortItems}
		autoHighlight={autoHighlight}
		inputProps={inputProps}
		onMenuVisibilityChange={onMenuVisibilityChange}
		onSelect={onSelect}
		renderMenu={renderMenu}
		wrapperProps={wrapperProps}
		wrapperStyle={wrapperStyle}
		open={open}
		getCaptionFromValue={getCaptionFromValue}
		getValueFromCaption={getValueFromCaption}
		renderItemGenerator={renderItemGenerator || defaults.renderItemGenerator} // will be used as the default for renderItem if none is supplied
	>
		{children}
	</Field>
);

AOAutocomplete.defaultProps = {
	getItemValue: defaults.getItemValue,
	shouldItemRender: defaults.shouldItemRender,
	sortItems: defaults.sortItems,
	renderMenu: defaults.renderMenu,
	onBlur: () => {},
	onChange: () => {},
	onDragStart: () => {},
	onDrop: () => {},
	onFocus: () => {},
	rightLabel: '',
	transition: false,
	placeholder: '',
	widthClassName: 'ao-width-M',
	additionalClassNames: '',
	onSelect: () => {},
	getCaptionFromValue: defaults.getCaptionFromValue,
	getValueFromCaption: defaults.getValueFromCaption,
};

AOAutocomplete.propTypes = {
	/** unique id for the input field */
	id: PropTypes.string.isRequired,

	/** name of the input field - should match the field name in the redux store */
	name: PropTypes.string.isRequired,

	/**
	 * Function for getting the display value from the items list. The default assumes the items has a 'value' property to use.
	 */
	getItemValue: PropTypes.func,

	/**
	 * Array of drop-down options. Many default functions assume an array of objects with a 'caption' and 'value' property.
	 */
	items: PropTypes.arrayOf(PropTypes.any).isRequired,

	/**
	 * Invoked for each entry in items. The default assumes the items have a 'caption' property to display.
	 * @see https://github.com/reactjs/react-autocomplete
	 */
	renderItem: PropTypes.func,

	/**
	 * Higher order function to create the renderItem function. This library passes renderItemGenerator the input value for use
	 */
	renderItemGenerator: PropTypes.func,

	/**
	 * Function to filter which items should render. The default assumes a 'caption' and 'value' property and will filter based on the input value being contained in either of those.
	 */
	shouldItemRender: PropTypes.func,

	/**
	 * Function to sort the items. The default assumes the items have a 'caption' property and sorts based on the location of the match.
	 * @see https://github.com/reactjs/react-autocomplete/blob/master/lib/utils.js
	 */
	sortItems: PropTypes.func,

	/**
	* ao-width size for the field (ao-width-XXS, ao-width-XS, ao-width-S, ao-width-M, ao-width-L, ao-width-XL, ao-width-XXL)
	* @see https://wordpress.aoins.com/AOUI/size-utilities/
	*/
	widthClassName: PropTypes.string,

	/** css class(es) to be appended to the input field */
	additionalClassNames: PropTypes.string,

	/** text to display as the label on the right of the input, often used for checkboxes and radio inputs */
	rightLabel: PropTypes.string,

	/** set this to true if you want to use transitions when a field becomes visible */
	transition: PropTypes.bool,

	/** input text placeholder */
	placeholder: PropTypes.string,

	/** Error validation function */
	validate: PropTypes.oneOfType([PropTypes.func, PropTypes.arrayOf(PropTypes.func)]),

	/** Warning validation function */
	warn: PropTypes.oneOfType([PropTypes.func, PropTypes.arrayOf(PropTypes.func)]),

	/** onBlur handler */
	onBlur: PropTypes.func,

	/** onChange handler */
	onChange: PropTypes.func,

	/** onDragStart handler */
	onDragStart: PropTypes.func,

	/** onDrop handler */
	onDrop: PropTypes.func,

	/** onFocus handler */
	onFocus: PropTypes.func,

	/** the options to display in the select */
	children: PropTypes.node,

	/**
	 * Additional react-autocomplete props
	 * @see https://github.com/reactjs/react-autocomplete
	 */

	/** Specifiy whether or not to automatically highlight the top match in the dropdown */
	autoHighlight: PropTypes.bool,

	/** Props applied to the input element */
	inputProps: PropTypes.objectOf(PropTypes.any),

	/**
	 * Invoked when the dropdown menu's visibility changes. Arguments: isOpen: Boolean
	 */
	onMenuVisibilityChange: PropTypes.func,

	/**
	 * Invoked when the user selects an item. Arguments: value: String, item: Any
	 */
	onSelect: PropTypes.func,

	/**
	 * Invoked to generate the render tree for the dropdown menu. Arguments: items: Array<Any>, value: String, styles: Object
	 */
	renderMenu: PropTypes.func,

	/** Props applied to element wrapping the input and drowdown menu */
	wrapperProps: PropTypes.objectOf(PropTypes.any),

	/** Specify styles for element wrapping the input and drowdown menu */
	wrapperStyle: PropTypes.objectOf(PropTypes.any),

	/** Function to get an item caption based on the value */
	getCaptionFromValue: PropTypes.func,

	/** Function to get an item value based on the caption */
	getValueFromCaption: PropTypes.func,

	/** True when the suggestions for the autocomplete should show */
	open: PropTypes.bool,
};

export default AOAutocomplete;
