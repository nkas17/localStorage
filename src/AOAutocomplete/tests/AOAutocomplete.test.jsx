import React from 'react';
import { reduxForm, reducer as formReducer } from 'redux-form';
import { createStore, combineReducers } from 'redux';
import { Provider } from 'react-redux';
import { snapshotRunner } from 'ao-react-redux-test-util';
import AOAutocomplete from '../AOAutocomplete';

// Set up some default test data
const defaultProps = {
	id: 'test',
	name: 'test',
	items: [
		{ caption: 'Caption1', value: 'Value1' },
		{ caption: 'Caption2', value: 'Value2' },
	],
};

const allProps = {
	id: 'test',
	name: 'test',
	items: [
		{ customCaption: 'Caption1', customValue: 'Value1' },
		{ customCaption: 'Caption2', customValue: 'Value2' },
	],
	getItemValue: item => item.customValue,
	renderItem: item => <span>{item.customCaption}</span>,
	shouldItemRender: () => true,
	sortItems: a => (a.customCaption === 'Caption1' ? -1 : 1),
	onBlur: () => {},
	onChange: () => {},
	onDragStart: () => {},
	onDrop: () => {},
	onFocus: () => {},
	widthClassName: 'some-width-class',
	additionalClassName: 'some-additional-class',
	rightLabel: 'probably unused',
	transition: false,
	placeholder: 'placeholder',
	validate: () => true,
	warn: () => true,
	children: '<div />',
	autoHighlight: false,
	inputProps: { extraProp: 'extraProp' },
	menuStyle: { background: 'black' },
	onMenuVisibilityChange: () => {},
	onSelect: () => {},
	renderMenu: items => <div>{items}</div>,
	wrapperProps: { wrapperProp: 'wrapperProp' },
	wrapperStyle: { background: 'white' },
};

// Set up the specific test scenarios
const scenarios = {
	'renders AOAutocomplete correctly with defaults': defaultProps,
	'renders AOAutocomplete correctly with all props': allProps,
	'renders AOAutocomplete correctly with open as false': Object.assign({}, defaultProps, {
		open: false,
	}),
};

const AOAutocompleteWithRedux = reduxForm({
	form: 'AOAutocompleteForm',
})(AOAutocomplete);

const store = createStore(combineReducers({ form: formReducer }));

snapshotRunner(scenarios, props => <Provider store={store}><AOAutocompleteWithRedux {...props} /></Provider>);
