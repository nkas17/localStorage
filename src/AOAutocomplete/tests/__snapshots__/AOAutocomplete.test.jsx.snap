// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders AOAutocomplete correctly with all props 1`] = `
<span>
  <span>
    <div
      style={
        Object {
          "background": "white",
        }
      }
      wrapperProp="wrapperProp"
    >
      <input
        aria-autocomplete="list"
        aria-expanded={false}
        autoComplete="off"
        className="ao-textbox some-width-class   "
        extraProp="extraProp"
        id="test"
        name="test"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="placeholder"
        role="combobox"
        value=""
      />
    </div>
  </span>
</span>
`;

exports[`renders AOAutocomplete correctly with defaults 1`] = `
<span>
  <span>
    <div
      style={
        Object {
          "display": "inline-block",
        }
      }
    >
      <input
        aria-autocomplete="list"
        aria-expanded={false}
        autoComplete="off"
        className="ao-textbox ao-width-M   "
        id="test"
        name="test"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder=""
        role="combobox"
        value=""
      />
    </div>
  </span>
</span>
`;

exports[`renders AOAutocomplete correctly with open as false 1`] = `
<span>
  <span>
    <div
      style={
        Object {
          "display": "inline-block",
        }
      }
    >
      <input
        aria-autocomplete="list"
        aria-expanded={false}
        autoComplete="off"
        className="ao-textbox ao-width-M   "
        id="test"
        name="test"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder=""
        role="combobox"
        value=""
      />
    </div>
  </span>
</span>
`;
